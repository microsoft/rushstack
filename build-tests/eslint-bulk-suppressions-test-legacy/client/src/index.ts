// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/* Top-level scope code samples */
// scopeId: '.'
let exampleString: string = 5 + '';

const exampleObject = {
  exampleString: exampleString
};

/* Function scope code samples */
export function exampleFunction() {
  const {}: Object = exampleObject;

  // scopeId: '.exampleFunction'
  !!!exampleString as Boolean;
}

// scope: '.ArrowFunctionExpression',
export const x = () => {},
  // scopeId: '.y'
  y = () => {},
  // scopeId: '.z'
  z = () => {};

/* Class scope code samples */
export class ExampleClass {
  // scopeId: '.ExampleClass'
  exampleClassProperty: String = exampleString + '4';

  exampleMethod() {
    // scopeId: '.exampleClass.exampleMethod'
    var exampleVar;
    return exampleVar;
  }
}

/* Variable and anonymous constructs code samples */
export const exampleArrowFunction = () => {
  const exampleBoolean = true;
  if (exampleBoolean) {
  }

  exampleObject['exampleString'];
};

export const exampleAnonymousClass = class {
  exampleClassProperty = 'x' + 'y';

  // scopeId: '.exampleAnonymousClass.constructor'
  constructor() {}

  set exampleSetGet(val: string) {
    // scopeId: '.exampleAnonymousClass.exampleSetGet'
    let exampleVariable: Number = 1;
    this.exampleClassProperty = val + exampleVariable;
  }

  get exampleSetGet() {
    // scopeId: '.exampleAnonymousClass.exampleSetGet'
    return this.exampleClassProperty as String as string;
  }
};
