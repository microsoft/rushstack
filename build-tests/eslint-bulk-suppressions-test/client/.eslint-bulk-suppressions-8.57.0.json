{
  "suppressions": [
    {
      "file": "src/index.ts",
      "scopeId": ".",
      "rule": "prefer-const"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".ExampleClass",
      "rule": "@typescript-eslint/ban-types"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".ExampleClass",
      "rule": "@typescript-eslint/explicit-member-accessibility"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".ExampleClass.exampleMethod",
      "rule": "@typescript-eslint/explicit-function-return-type"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".ExampleClass.exampleMethod",
      "rule": "@typescript-eslint/explicit-member-accessibility"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".ExampleClass.exampleMethod",
      "rule": "@typescript-eslint/typedef"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".ExampleClass.exampleMethod",
      "rule": "no-var"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleAnonymousClass",
      "rule": "@typescript-eslint/explicit-member-accessibility"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleAnonymousClass",
      "rule": "@typescript-eslint/typedef"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleAnonymousClass",
      "rule": "no-useless-concat"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleAnonymousClass.constructor",
      "rule": "@typescript-eslint/explicit-member-accessibility"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleAnonymousClass.exampleSetGet",
      "rule": "@typescript-eslint/ban-types"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleAnonymousClass.exampleSetGet",
      "rule": "@typescript-eslint/explicit-function-return-type"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleAnonymousClass.exampleSetGet",
      "rule": "@typescript-eslint/explicit-member-accessibility"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleArrowFunction",
      "rule": "@typescript-eslint/explicit-function-return-type"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleArrowFunction",
      "rule": "@typescript-eslint/typedef"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleArrowFunction",
      "rule": "dot-notation"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleArrowFunction",
      "rule": "no-empty"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleArrowFunction",
      "rule": "no-unused-expressions"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleFunction",
      "rule": "@typescript-eslint/ban-types"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleFunction",
      "rule": "@typescript-eslint/explicit-function-return-type"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleFunction",
      "rule": "no-empty-pattern"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleFunction",
      "rule": "no-extra-boolean-cast"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".exampleObject",
      "rule": "@typescript-eslint/typedef"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".x",
      "rule": "@typescript-eslint/explicit-function-return-type"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".y",
      "rule": "@typescript-eslint/explicit-function-return-type"
    },
    {
      "file": "src/index.ts",
      "scopeId": ".z",
      "rule": "@typescript-eslint/explicit-function-return-type"
    }
  ]
}
