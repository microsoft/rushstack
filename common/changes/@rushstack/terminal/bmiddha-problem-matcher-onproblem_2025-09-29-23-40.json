{
  "changes": [
    {
      "packageName": "@rushstack/terminal",
      "comment": "Add ProblemCollector.onProblem notification callback",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/terminal"
}