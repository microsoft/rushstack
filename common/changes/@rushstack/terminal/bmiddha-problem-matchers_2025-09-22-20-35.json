{
  "changes": [
    {
      "packageName": "@rushstack/terminal",
      "comment": "Add `ProblemCollector extends TerminalWritable` API which matches and collects VS Code style problem matchers",
      "type": "patch"
    }
  ],
  "packageName": "@rushstack/terminal"
}