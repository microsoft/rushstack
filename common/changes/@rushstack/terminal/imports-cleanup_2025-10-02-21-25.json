{
  "changes": [
    {
      "comment": "Normalize import of builtin modules to use the `node:` protocol.",
      "type": "minor",
      "packageName": "@rushstack/terminal"
    }
  ],
  "packageName": "@rushstack/terminal",
  "email": "iclanton@users.noreply.github.com"
}
