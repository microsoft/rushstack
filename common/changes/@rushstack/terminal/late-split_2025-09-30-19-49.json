{
  "changes": [
    {
      "packageName": "@rushstack/terminal",
      "comment": "Update API contract for `SplitterTransform` to support adding and removing destinations after creation.",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/terminal"
}