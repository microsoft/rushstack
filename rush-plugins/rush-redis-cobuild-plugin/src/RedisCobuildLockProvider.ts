// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { createClient } from '@redis/client';
import type {
  RedisClientOptions,
  RedisClientType,
  RedisFunctions,
  RedisModules,
  RedisScripts
} from '@redis/client';

import type {
  ICobuildLockProvider,
  ICobuildContext,
  ICobuildCompletedState,
  RushSession
} from '@rushstack/rush-sdk';
import type { ITerminal } from '@rushstack/terminal';

/**
 * The redis client options
 * @beta
 */
export interface IRedisCobuildLockProviderOptions extends RedisClientOptions {
  /**
   * The environment variable name for the redis password
   */
  passwordEnvironmentVariable?: string;
}

const COMPLETED_STATE_SEPARATOR: ';' = ';';

/**
 * @beta
 */
export class RedisCobuildLockProvider implements ICobuildLockProvider {
  private readonly _options: IRedisCobuildLockProviderOptions;
  private readonly _terminal: ITerminal;
  private readonly _lockKeyIdentifierMap: WeakMap<ICobuildContext, string> = new WeakMap<
    ICobuildContext,
    string
  >();
  private readonly _completedStateKeyIdentifierMap: WeakMap<ICobuildContext, string> = new WeakMap<
    ICobuildContext,
    string
  >();

  private readonly _redisClient: RedisClientType<RedisModules, RedisFunctions, RedisScripts, 2 | 3>;

  public constructor(options: IRedisCobuildLockProviderOptions, rushSession: RushSession) {
    this._options = RedisCobuildLockProvider.expandOptionsWithEnvironmentVariables(options);
    // Provide a default reconnect strategy that prevents more than 5 reconnect attempts.
    this._options.socket = {
      reconnectStrategy: (count: number) => {
        this._terminal.writeErrorLine(`Redis client reconnecting attempt #${count}`);
        return count < 5 ? count * 1000 : false;
      },
      ...this._options.socket
    };
    this._terminal = rushSession.getLogger('RedisCobuildLockProvider').terminal;
    try {
      this._redisClient = createClient(this._options);
    } catch (e) {
      throw new Error(`Failed to create redis client: ${e.message}`);
    }
  }

  public static expandOptionsWithEnvironmentVariables(
    options: IRedisCobuildLockProviderOptions,
    environment: NodeJS.ProcessEnv = process.env
  ): IRedisCobuildLockProviderOptions {
    const finalOptions: IRedisCobuildLockProviderOptions = { ...options };
    const missingEnvironmentVariables: Set<string> = new Set<string>();

    if (finalOptions.passwordEnvironmentVariable) {
      const password: string | undefined = environment[finalOptions.passwordEnvironmentVariable];
      if (password !== undefined) {
        finalOptions.password = password;
      } else {
        missingEnvironmentVariables.add(finalOptions.passwordEnvironmentVariable);
      }
      finalOptions.passwordEnvironmentVariable = undefined;
    }

    if (missingEnvironmentVariables.size) {
      throw new Error(
        `The "RedisCobuildLockProvider" tries to access missing environment variable${
          missingEnvironmentVariables.size > 1 ? 's' : ''
        }: ${Array.from(missingEnvironmentVariables).join(
          ', '
        )}\nPlease check the configuration in rush-redis-cobuild-plugin.json file`
      );
    }
    return finalOptions;
  }

  public async connectAsync(): Promise<void> {
    try {
      await this._redisClient.connect();
      // Check the connection works at early stage
      await this._redisClient.ping();
    } catch (e) {
      throw new Error(`Failed to connect to redis server: ${e.message}`);
    }

    // Register error event handler to avoid process exit when redis client error occurs.
    this._redisClient.on('error', (e: Error) => {
      if (e.message) {
        this._terminal.writeErrorLine(`Redis client error: ${e.message}`);
      } else {
        this._terminal.writeErrorLine(`Redis client error: ${e}`);
      }
    });
  }

  public async disconnectAsync(): Promise<void> {
    try {
      await this._redisClient.destroy();
    } catch (e) {
      throw new Error(`Failed to disconnect to redis server: ${e.message}`);
    }
  }

  /**
   * Acquiring the lock based on the specific context.
   *
   * NOTE: this is a reentrant lock implementation
   */
  public async acquireLockAsync(context: ICobuildContext): Promise<boolean> {
    const { _terminal: terminal } = this;
    const { lockKey, lockExpireTimeInSeconds, runnerId } = context;
    let result: boolean = false;
    const lockKeyIdentifier: string = this._getLockKeyIdentifier(context);
    try {
      // According to the doc, the reply of set command is either "OK" or nil. The reply doesn't matter
      await this._redisClient.set(lockKey, runnerId, {
        NX: true,
        // call EXPIRE in an atomic command
        EX: lockExpireTimeInSeconds
        // Do not specify GET here since using NX ane GET together requires Redis@7.
      });
      // Just read the value by lock key to see wether it equals current runner id
      const value: string | null = await this._redisClient.get(lockKey);
      if (value === null) {
        // This should not happen.
        throw new Error(`Get redis key failed: ${lockKey}`);
      }
      result = value === runnerId;
      if (result) {
        terminal.writeDebugLine(
          `Successfully acquired ${lockKeyIdentifier} to runner(${runnerId}) and it expires in ${lockExpireTimeInSeconds}s`
        );
      } else {
        terminal.writeDebugLine(`Failed to acquire ${lockKeyIdentifier}, locked by runner ${value}`);
      }
    } catch (e) {
      throw new Error(`Error occurs when acquiring ${lockKeyIdentifier}: ${e.message}`);
    }
    return result;
  }

  public async renewLockAsync(context: ICobuildContext): Promise<void> {
    const { _terminal: terminal } = this;
    const { lockKey, lockExpireTimeInSeconds } = context;
    const lockKeyIdentifier: string = this._getLockKeyIdentifier(context);
    try {
      await this._redisClient.expire(lockKey, lockExpireTimeInSeconds);
    } catch (e) {
      throw new Error(`Failed to renew ${lockKeyIdentifier}: ${e.message}`);
    }
    terminal.writeDebugLine(`Renewed ${lockKeyIdentifier} expires in ${lockExpireTimeInSeconds} seconds`);
  }

  public async setCompletedStateAsync(
    context: ICobuildContext,
    state: ICobuildCompletedState
  ): Promise<void> {
    const { _terminal: terminal } = this;
    const { completedStateKey: key } = context;
    const value: string = this._serializeCompletedState(state);
    const completedStateKeyIdentifier: string = this._getCompletedStateKeyIdentifier(context);
    try {
      await this._redisClient.set(key, value);
    } catch (e) {
      throw new Error(`Failed to set ${completedStateKeyIdentifier}: ${e.message}`);
    }
    terminal.writeDebugLine(`Set ${completedStateKeyIdentifier}: ${value}`);
  }

  public async getCompletedStateAsync(context: ICobuildContext): Promise<ICobuildCompletedState | undefined> {
    const { _terminal: terminal } = this;
    const { completedStateKey: key } = context;
    const completedStateKeyIdentifier: string = this._getCompletedStateKeyIdentifier(context);
    let state: ICobuildCompletedState | undefined;
    try {
      const value: string | null = await this._redisClient.get(key);
      if (value) {
        state = this._deserializeCompletedState(value);
      }
      terminal.writeDebugLine(`Get ${completedStateKeyIdentifier}: ${value}`);
    } catch (e) {
      throw new Error(`Failed to get ${completedStateKeyIdentifier}: ${e.message}`);
    }
    return state;
  }

  private _serializeCompletedState(state: ICobuildCompletedState): string {
    // Example: SUCCESS;1234567890
    // Example: FAILURE;1234567890
    const { status, cacheId } = state;
    return [status, cacheId].join(COMPLETED_STATE_SEPARATOR);
  }

  private _deserializeCompletedState(state: string): ICobuildCompletedState | undefined {
    const [status, cacheId] = state.split(COMPLETED_STATE_SEPARATOR);
    return { status: status as ICobuildCompletedState['status'], cacheId };
  }

  private _getLockKeyIdentifier(context: ICobuildContext): string {
    let lockKeyIdentifier: string | undefined = this._lockKeyIdentifierMap.get(context);
    if (lockKeyIdentifier === undefined) {
      const { lockKey, packageName, phaseName } = context;
      lockKeyIdentifier = `lock(${lockKey})_package(${packageName})_phase(${phaseName})`;
      this._lockKeyIdentifierMap.set(context, lockKeyIdentifier);
    }
    return lockKeyIdentifier;
  }

  private _getCompletedStateKeyIdentifier(context: ICobuildContext): string {
    let completedStateKeyIdentifier: string | undefined = this._completedStateKeyIdentifierMap.get(context);
    if (completedStateKeyIdentifier === undefined) {
      const { completedStateKey, packageName, phaseName } = context;
      completedStateKeyIdentifier = `completed_state(${completedStateKey})_package(${packageName})_phase(${phaseName})`;
      this._completedStateKeyIdentifierMap.set(context, completedStateKeyIdentifier);
    }
    return completedStateKeyIdentifier;
  }
}
