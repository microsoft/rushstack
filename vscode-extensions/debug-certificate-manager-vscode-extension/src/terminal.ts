// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { ITerminal } from '@rushstack/terminal';
import * as vscode from 'vscode';
import { stripVTControlCharacters } from 'node:util';

export async function runWorkspaceCommandAsync({
  terminalOptions,
  commandLine,
  terminal
}: {
  terminalOptions: vscode.TerminalOptions;
  commandLine: string;
  terminal: ITerminal;
}): Promise<string> {
  const vsTerminal: vscode.Terminal = vscode.window.createTerminal(terminalOptions);

  // wait for shell to bootup and vs code shell integration to kick-in
  const shellIntegration: vscode.TerminalShellIntegration =
    vsTerminal.shellIntegration ??
    (await new Promise((resolve, reject) => {
      const shellIntegrationDisposable: vscode.Disposable = vscode.window.onDidChangeTerminalShellIntegration(
        (event) => {
          if (event.terminal !== vsTerminal) {
            return;
          }

          resolve(event.shellIntegration);
          shellIntegrationDisposable?.dispose();
        }
      );
    }));

  // Run the command through shell integration and grab output
  return new Promise<string>((resolve, reject) => {
    let outputStream: string = '';

    // start output capturing with the start execution event
    const startExecutionDisposable: vscode.Disposable = vscode.window.onDidStartTerminalShellExecution(
      async (event) => {
        if (event.terminal !== vsTerminal) {
          return;
        }
        terminal.writeLine(`Terminal shell execution started`);

        for await (const chunk of event.execution.read()) {
          outputStream += chunk;
        }
      }
    );

    // collect output and exit code
    const endExecutionDisposable: vscode.Disposable = vscode.window.onDidEndTerminalShellExecution(
      (event) => {
        if (event.terminal !== vsTerminal) {
          return;
        }

        terminal.writeLine(`Terminal shell execution ended with exit code ${event.exitCode}`);
        outputStream = outputStream.trim();
        outputStream = stripVTControlCharacters(outputStream);
        terminal.writeLine(`Terminal output: ${outputStream}`);

        endExecutionDisposable.dispose();
        startExecutionDisposable.dispose();

        if (event.exitCode === 0) {
          resolve(outputStream);
        } else {
          reject(outputStream);
        }
      }
    );

    shellIntegration.executeCommand(commandLine);
    terminal.writeLine(`Executing command: ${commandLine}`);
  }).finally(() => {
    vsTerminal.dispose();
  });
}
