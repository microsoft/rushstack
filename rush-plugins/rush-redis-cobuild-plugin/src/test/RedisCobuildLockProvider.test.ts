// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { ConsoleTerminalProvider } from '@rushstack/terminal';
import * as redisAPI from '@redis/client';
import type { RedisClientType } from '@redis/client';

import {
  type ICobuildCompletedState,
  type ICobuildContext,
  OperationStatus,
  RushSession
} from '@rushstack/rush-sdk';
import { type IRedisCobuildLockProviderOptions, RedisCobuildLockProvider } from '../RedisCobuildLockProvider';

const rushSession: RushSession = new RushSession({
  terminalProvider: new ConsoleTerminalProvider(),
  getIsDebugMode: () => false
});

describe(RedisCobuildLockProvider.name, () => {
  let storage: Record<string, string> = {};
  beforeEach(() => {
    jest.spyOn(redisAPI, 'createClient').mockImplementation(() => {
      return {
        expire: jest.fn().mockResolvedValue(undefined),
        set: jest
          .fn()
          .mockImplementation((key: string, value: string, options?: { NX?: boolean; GET?: boolean }) => {
            // https://redis.io/commands/set/
            const oldValue: string | undefined = storage[key];
            const { NX, GET } = options || {};
            let didSet: boolean = false;
            if (NX) {
              if (!storage[key]) {
                storage[key] = value;
                didSet = true;
              }
            } else {
              storage[key] = value;
              didSet = true;
            }

            if (GET) {
              if (oldValue === undefined) {
                return null;
              } else {
                return oldValue;
              }
            } else {
              if (didSet) {
                return 'OK';
              } else {
                return null;
              }
            }
          }),
        get: jest.fn().mockImplementation((key: string) => {
          return storage[key];
        })
      } as unknown as RedisClientType;
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
    storage = {};
  });

  function prepareSubject(): RedisCobuildLockProvider {
    return new RedisCobuildLockProvider({} as IRedisCobuildLockProviderOptions, rushSession);
  }

  const context: ICobuildContext = {
    contextId: 'context_id',
    cacheId: 'cache_id',
    lockKey: 'lock_key',
    lockExpireTimeInSeconds: 30,
    completedStateKey: 'completed_state_key',
    clusterId: 'cluster_id',
    runnerId: 'runner_id',
    packageName: 'package_name',
    phaseName: 'phase_name'
  };

  it('expands options with environment variables', () => {
    const expectedOptions = {
      password: 'redis123' // [SuppressMessage("Microsoft.Security", "CS001:SecretInline", Justification="Password used in unit test.")]
    };
    const actualOptions = RedisCobuildLockProvider.expandOptionsWithEnvironmentVariables(
      {
        passwordEnvironmentVariable: 'REDIS_PASS'
      },
      {
        REDIS_PASS: 'redis123'
      }
    );
    expect(actualOptions).toEqual(expectedOptions);
  });

  it('throws error with missing environment variables', () => {
    expect(() => {
      RedisCobuildLockProvider.expandOptionsWithEnvironmentVariables(
        {
          passwordEnvironmentVariable: 'REDIS_PASS'
        },
        {}
      );
    }).toThrowErrorMatchingSnapshot();
  });

  it('acquires lock success', async () => {
    const subject: RedisCobuildLockProvider = prepareSubject();
    const result: boolean = await subject.acquireLockAsync(context);
    expect(result).toBe(true);
  });

  it('acquires lock is a reentrant lock', async () => {
    const subject: RedisCobuildLockProvider = prepareSubject();
    const result1: boolean = await subject.acquireLockAsync(context);
    expect(result1).toBe(true);
    const result2: boolean = await subject.acquireLockAsync(context);
    expect(result2).toBe(true);
  });

  it('acquires lock fails with a different runner', async () => {
    const subject: RedisCobuildLockProvider = prepareSubject();
    const result1: boolean = await subject.acquireLockAsync(context);
    expect(result1).toBe(true);
    const cobuildContext: ICobuildContext = {
      ...context,
      runnerId: 'other_runner_id'
    };
    const result2: boolean = await subject.acquireLockAsync(cobuildContext);
    expect(result2).toBe(false);
  });

  it('set and get completedState works', async () => {
    const subject: RedisCobuildLockProvider = prepareSubject();
    const cacheId: string = 'foo';
    const status: ICobuildCompletedState['status'] = OperationStatus.SuccessWithWarning;
    expect(() => subject.setCompletedStateAsync(context, { status, cacheId })).not.toThrowError();
    const actualState: ICobuildCompletedState | undefined = await subject.getCompletedStateAsync(context);
    expect(actualState?.cacheId).toBe(cacheId);
    expect(actualState?.status).toBe(status);
  });
});
