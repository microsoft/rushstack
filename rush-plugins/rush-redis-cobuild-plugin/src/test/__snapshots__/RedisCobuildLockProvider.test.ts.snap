// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RedisCobuildLockProvider throws error with missing environment variables 1`] = `
"The \\"RedisCobuildLockProvider\\" tries to access missing environment variable: REDIS_PASS
Please check the configuration in rush-redis-cobuild-plugin.json file"
`;
